### DISCUSSION TITLE
Vite.js + Eslint unable to resolve path alias

### DISCUSSION THREAD

**Discussion by Delicious-Bacon** [eslint/eslint#14667] [Q&A] [answered]
I use Vue 3 on Vite.js with Eslint + Airbnb config. Airbnb config has a rule `eslint(import/no-unresolved)`, which is good, but Eslint doesn't know how to resolve alias path.

I want to use aliases for paths — example: 
`import TableComponent from '@/components/table/TableComponent.vue'`

I managed to set up my `vite.config.js` so that the app can resolve paths like this:
```
import path from 'path';
import { defineConfig } from 'vite';

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [vue()],
  resolve: {
    alias: [{
        find: "@", replacement: path.resolve(__dirname, 'src')
      },],
  },
});
```
Vue app works like that and resolves the import path correctly, but Eslint keeps reporting the error:
`Unable to resolve path to module eslint(import/no-unresolved)`

How and where can I tell Eslint how to resolve aliases?

I have tried: `npm install eslint-plugin-import eslint-import-resolver-alias --save-dev`
```
// .eslintrc.js

extends: [
    'eslint:recommended',
    'plugin:import/recommended',
    'airbnb-base',
    'plugin:vue/vue3-strongly-recommended',
],

settings: {
    'import/resolver': {
      alias: {
        map: [
          ['@', 'src'],
        ],
      },
    },
  },
```
But that doesn't work.

**Comment by aladdin-add** [comment, marked as answer]
this is related to the `eslint-plugin-import`, which is not maintained by the eslint team. you can file the issue in that repo.

maybe you can try this: 
```js
settings: {
    'import/resolver': {
      alias: {
        map: [
          ['@', './src'],
        ],
      },
    },
  },
```

**Reply by Delicious-Bacon** [reply]
It works! Thank you so much! <3

It was driving me crazy. ^^;

**Reply by liquidvisual** [reply]
Doesn't work

**Reply by Delicious-Bacon** [reply]
It depends on your project's folder structure (and maybe how your OS resolves path).

**Reply by floroz** [reply]
This worked for me in a monorepo setup with PNPM Workspaces and Lerna.

Make sure you've installed the `eslint-import-resolver-typescript`

```js
  settings: {
    'import/resolver': {
      typescript: {},
      alias: {
        map: [
          ['@', './src'],
        ],
      },
  },
```

**Reply by Vercadium** [reply]
Just to be 💯 clear for those that landed on this suggested answer after a search and 👎

These configuration snippets will **only** work if you have an appropriate resolver installed - you **cannot** currently accommodate the alias with `eslint` and `eslint-plugin-import` alone.

Install `eslint-import-resolver-alias` then add the following to your ESLint configuration (note the `map` can be omitted):

```js
settings: {
  'import/resolver': {
    alias: [
      ['@', './src'],
    ],
  },
},
  ```

**Reply by CyanSalt** [reply]
> This worked for me in a monorepo setup with PNPM Workspaces and Lerna.
> 
> Make sure you've installed the `eslint-import-resolver-typescript`
> 
> ```js
>   settings: {
>     'import/resolver': {
>       typescript: {},
>       alias: {
>         map: [
>           ['@', './src'],
>         ],
>       },
>   },
> ```

For those of you who, like me, have tried to use multiple resolvers and it didn't work, please note that the order here must be consistent, `typescript` needs to come before `alias`.

**Reply by vincerubinetti** [reply]
It's odd that (as of writing) Vite scaffolds a React project that has the import plugin, but not the resolver it needs...

Rather than installing yet another package and having more config, I think I'm going to just remove the plugin and rely on Typescript or `vite build` to tell me if there's an import problem.

**Reply by nopeless** [reply]
@CyanSalt That is an insane detail that I would have never caught. Thank you

**Reply by prakhar-garg-dev** [reply]
Hey I was working with eslint.cjs files and used eslint-plugin-import and eslint-plugin-import-alias together on windows operating system
I did the following to fix the issue 

```js

const path = require('path'); // requiring the path as the file is commonjs
module.exports = {
  root: true,
  env: { browser: true, es2020: true },
  extends: ['eslint:recommended', 'plugin:import/recommended', 'plugin:react/recommended', 'plugin:react/jsx-runtime', 'plugin:react-hooks/recommended', 'prettier', 'plugin:storybook/recommended'],
  
 
  ignorePatterns: ['dist', '.eslintrc.cjs'],
  parserOptions: { ecmaVersion: 'latest', sourceType: 'module' },
  settings: {
    react: { version: '18.2' },
     'import/resolver': {  
       node: {
         extensions: ['.js', '.jsx'],
       },
       alias: [
         ['~', path.resolve(__dirname, './src')],// used the path module to resolve the path here 
         ['views',path.resolve(__dirname,'./src/views')]
       ],
        
     
    },
  },
  plugins: ['react-refresh'],
  rules: {
    'react-refresh/only-export-components': [
      'warn',
      { allowConstantExport: true },
    ],
  },
}
```
Similarly in vite config I did the following :
```js
import { defineConfig } from 'vite'
import react from '@vitejs/plugin-react'
import eslint from 'vite-plugin-eslint';
// To get the resolution from the node path module for the file system
import path from 'node:path';
import { fileURLToPath } from 'url';
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
export default defineConfig({
  plugins: [react(), { 
// do not fail on serve (i.e. local development)
    ...eslint({
      failOnWarning: false,
      failOnError: false,
    }),
    apply: 'serve',
    enforce: 'post'
  }],
  // Used to resolve alias in absolute paths in vite
  resolve: {
    alias: [
      { find: '~', replacement: path.resolve(__dirname,'./src'), },
      { find: 'components', replacement: path.resolve(__dirname, './src/components') },
      { find: 'views', replacement: path.resolve(__dirname, './src/views') },
    ]
  },
  
})
```

**Reply by Doug-Colin** [reply]
### If anyone is still having trouble, despite trying the other solutions, the only way I could get it to work was this:

- Install plugins ` eslint-import-resolver-alias` , and `eslint-import-resolver-typescript` ([1](https://github.com/johvin/eslint-import-resolver-alias#readme), [2](https://github.com/import-js/eslint-import-resolver-typescript#readme)), as @CyanSalt and @Prakhargarg-2010196  mentioned.
- Add 'import-resolver-alias' to the `plugins` array. 

### Notes that may help you:
- I'm using Vite as well (on Windows 10).
- I'm using Typescript so if you're not, you shouldn't need `eslint-import-resolver-typescript`.
- My path alias is defined in both my viteconfig.ts & tsconfig.json files, included below.
- Unlike @CyanSalt, I did not need to add the alias to my 'import/resolver' settings, but if you're not using `eslint-import-resolver-typescript`, you probably need to (that plugin can access the alias in your tsconfig.json).
- If you're using  `eslint-import-resolver-typescript`, I received the following error when trying to lint all TypeScript files via  `npx eslint "./src/**/*.{ts,tsx}" --fix`, which I resolved heavy-handedly by deleting package.json and reinstalling dependencies via npm install:
```Oops! Something went wrong! :(
ESLint: 8.49.0
ESLint couldn't find the plugin "eslint-plugin-eslint-import-resolver-alias".
(The package "eslint-plugin-eslint-import-resolver-alias" was not found when loaded as a Node module from the directory ...
```
- My versions of relevant plugins/dependencies:
```js
    "@typescript-eslint/eslint-plugin": "^6.0.0",
    "@typescript-eslint/parser": "^6.0.0",
    "eslint": "^8.49.0",
    "eslint-config-prettier": "^9.0.0",
    "eslint-import-resolver-alias": "^1.1.2",
    "eslint-import-resolver-typescript": "^3.6.0",
    "eslint-plugin-import": "^2.28.1",
    "eslint-plugin-prettier": "^5.0.0",
    "prettier": "^3.0.3",
    "typescript": "^5.0.2",
    "vite": "^4.4.5"
```
 
- eslintrc.cjs:
```js
module.exports = {
  root: true,
  env: { browser: true, es2020: true },
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:react-hooks/recommended',
    'plugin:prettier/recommended',
    'plugin:import/errors',
    'plugin:import/warnings',
    'plugin:import/typescript',
  ],
  ignorePatterns: ['dist', '.eslintrc.cjs'],
  parser: '@typescript-eslint/parser',
  plugins: [
    'import',
     /* This line here was key */
    'import-resolver-alias', 
    'react-refresh',
  ],
  settings: {
    'import/parsers': {
      '@typescript-eslint/parser': ['.ts', '.tsx'],
    },
    'import/resolver': {
      typescript: {
        alwaysTryTypes: true,
      },
    },
  },
  rules: {
    'import/no-unresolved': 'error',
    'react-refresh/only-export-components': [
      'warn',
      { allowConstantExport: true },
    ],
  },
}
```

- tsconfig.json:
```js
{
  "compilerOptions": {
    "target": "ES2020",
    "useDefineForClassFields": true,
    "lib": [
      "ES2020",
      "DOM",
      "DOM.Iterable"
    ],
    "module": "ESNext",
    "skipLibCheck": true,
     /* Resolve paths/path alias without error*/
    "baseUrl": ".",
    "paths": {          
        "./src/*"
      ]
    },
    /* Bundler mode */
    "moduleResolution": "bundler",
    "allowImportingTsExtensions": true,
    "resolveJsonModule": true,
    "isolatedModules": true,
    "noEmit": true,
    "jsx": "react-jsx",
    /* Linting */
    "strict": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noFallthroughCasesInSwitch": true
  },
  "include": [
    "src"
, "project_notes/DialogDemo.tsx"  ],
  "references": [
    {
      "path": "./tsconfig.node.json"
    }
  ]
}
```

- viteconfig.ts:
```ts
import path from "path"
import { defineConfig } from "vite"

export default defineConfig({
  /* Resolve path alias */
  resolve: {
    alias: {                                                    
      "@": path.resolve(__dirname, "./src"),
    },
  },
  server: {
    proxy: {
         /* My proxy config*/
    }
  }
});
````

**Reply by prakhar-garg-dev** [reply]
I  think the inclusion of path should be somewhat resolved differently as it cause some major problems. Like its not utility that for each thing we need to install new plugin..

**Reply by Doug-Colin** [reply]
What do you mean by differently- that the import plugin would be better off approaching things differently, or that my solution would be better off approaching things differently? 

What major problems could this lead to?  

I'm not disputing you, I'm genuinely curious as this was my first time configuring eslint for TypeScript in Vite.

**Reply by prakhar-garg-dev** [reply]
@Doug-Colin  I am not saying that any of the answers is wrong but just suggesting that paths must be imported and handled in different way in vite itself like many things in configuring vite with other tools sometimes feels overhead with documentation leading to few hints to solve issues. Apologies if I said something wrong. 

And about problems ,
One of the problem I am still facing after solution I have written was that vite which I think should understand .jsx or js files by itself if extension is not provided is requiring me to write extension(even when I tried fixing it through eslint plugin import/extensions too. Like many little problems occur which are not traceable back to anything and are ambigous.

**Reply by mfalthaw** [reply]
If you're still having trouble and none of the other solutions worked for you, here's what did it for me

**.eslintrc.cjs**
```js
 'import/resolver': {
      node: {
        paths: ['src'],
        extensions: ['.js', '.jsx', '.ts', '.d.ts', '.tsx'],
      },
    },
```

**Reply by jdtorregrosas-circulee** [reply]
For me the only change needed was to include this in the plugins array:
```
  "plugins": ["import-resolver-alias"],
 ```
 this library `eslint-import-resolver-alias` did the whole trick, I didn't need this one `eslint-import-resolver-typescript`
 I'm using vite, typescript and react, oviously I added the aliases in the ts-config and vite-config

**Reply by ajaybandi14** [reply]
@jdtorregrosas-circulee This worked for me too.

**Reply by ianpaschal** [reply]
@jdtorregrosas-circulee are you sure that's the right name?

Adding the package you mentioned and that plugin name results in:

```
Oops! Something went wrong! :(
ESLint: 8.52.0
ESLint couldn't find the plugin "eslint-plugin-import-resolver-alias".
```

**Reply by prakhar-garg-dev** [reply]
> @jdtorregrosas-circulee are you sure that's the right name?
> 
> Adding the package you mentioned and that plugin name results in:
> 
> ```
> Oops! Something went wrong! :(
> ESLint: 8.52.0
> ESLint couldn't find the plugin "eslint-plugin-import-resolver-alias".
> ```

What you have written in eslint config ?

**Reply by ianpaschal** [reply]
`Package.json`
```
"eslint-import-resolver-alias": "^1.1.2",
```

`.eslintrc.cjs`
```
module.exports = {
  // ...
  plugins: [
    'import',
    'simple-import-sort',
    'import-newlines',
    'react-refresh',
    'import-resolver-alias',
  ],
  // ...
};
```

**Reply by prakhar-garg-dev** [reply]
Uhm Its correct according to me. Can you go through this repos' config and compare to yours if you are using [plain javascript](https://github.com/Prakhargarg-2010196/template-plain-vite) or see above replies of mine if you wamt to see the typescript one.

**Reply by ianpaschal** [reply]
Mine doesn't really look like yours. More like what @Doug-Colin shared (thanks!). I'm actually surprised to see your in repo that you don't have the plugin listed in the plugins array at all. I'm running into the same problem that Doug did, although I'd prefer not to resolve it by deleting package.json and manually reinstalling everything.

**Reply by ianpaschal** [reply]
If I remove the plugin from the plugins array and use the settings section to add the alias (similar to what you did), ESLlint still complains about lines like:

```ts
import { App } from '~/App';
```

Although annoyingly the error is 

> Missing file extension for "~/App"eslint[import/extensions](https://github.com/import-js/eslint-plugin-import/blob/v2.29.0/docs/rules/extensions.md)

 and if I add `.tsx`, the error becomes

> Unexpected use of file extension "tsx" for "~/App.tsx"eslint[import/extensions](https://github.com/import-js/eslint-plugin-import/blob/v2.29.0/docs/rules/extensions.md)

I can silence these by turning off the `import/extensions` rule but I still want it to force the use of extensions for everything that isn't a TS module. I can also configure the rule to give that behavior but that feels like it shouldn't be necessary since that's the default behavior of `import/extensions` in the first place if aliases aren't used. Somehow the alias plugin and the import plugin are not playing nice together.

**Reply by prakhar-garg-dev** [reply]
> If I remove the plugin from the plugins array and use the settings section to add the alias (similar to what you did), ESLlint still complains about lines like:
> 
> ```ts
> import { App } from '~/App';
> ```
> 
> Although annoyingly the error is 
> 
> > Missing file extension for "~/App"eslint[import/extensions](https://github.com/import-js/eslint-plugin-import/blob/v2.29.0/docs/rules/extensions.md)
> 
>  and if I add `.tsx`, the error becomes
> 
> > Unexpected use of file extension "tsx" for "~/App.tsx"eslint[import/extensions](https://github.com/import-js/eslint-plugin-import/blob/v2.29.0/docs/rules/extensions.md)
> 
> I can silence these by turning off the `import/extensions` rule but I still want it to force the use of extensions for everything that isn't a TS module. I can also configure the rule to give that behavior but that feels like it shouldn't be necessary since that's the default behavior of `import/extensions` in the first place if aliases aren't used. Somehow the alias plugin and the import plugin are not playing nice together.

Thats what i said in the above reply to @Doug-Colin  that the default functionality for this must be improved in some way. Btw try adding extensions to your paths ending like instead of ~/App  do App.jsx although ik it feels weird. Also restart app and then tell what you are getting now 

Moreover there was no need for me to define this  plugin in the plugins as to the documentation suggests this. ![IMG_20231104_193509.jpg](https://github.com/eslint/eslint/assets/77922738/b0f6382f-5ade-4acc-aaca-cd04ae181fe6)


Refer these: 

https://www.npmjs.com/package/eslint-plugin-import#resolvers
 https://www.npmjs.com/package/eslint-import-resolver-alias

**Reply by wmonecke** [reply]
@ianpaschal I am having the same issue with no avail.

**Reply by wmonecke** [reply]
@ianpaschal Nevermind let me save your day:

Try this:

```
// .eslintrc.json
       settings: {
    'import/resolver': {
      node: {
        paths: ['src'],
        extensions: ['.js', '.jsx', '.ts', '.d.ts', '.tsx'],
      },
      typescript: {
        project: './tsconfig.json',
      },
      alias: {
        map: [['~', path.resolve(__dirname, './src')]],
        extensions: ['.js', '.jsx', '.ts', '.d.ts', '.tsx'],
      },
    },
  },
```

**Reply by shennan** [reply]
Something I needed to do in VS Code after making changes was get the menu up (cmd + shift + p) and then `ESLint: Restart ESLint Server`.

**Reply by kyumoon** [reply]
`path.resolve` works

**Comment by daiwanxing** [comment]
after eslint 9 came out. i found an another workaound:

```js
// eslint.config.js
import viteConfig from "./vite.config.js";

//  make sure installed both latest eslint-import-resolver-vite and eslint-plugin-import.

export default [
    {
        settings: {
            "import/resolver": {
                typescript: true,
                node: true,
                vite: {
                  // the `eslint-import-resolver-vite` need to read your vite alias configuration and resolve it.
                    viteConfig: viteConfig,
                },
            },
        },
    }
]
```

**Comment by VGontier-cmd** [comment]
None of the solution worked for me with the new flat config introduced.
The only way I had to make it work is by using the `@babel/eslint-parser` and `@babel/preset-react` libs to parse it.
I also needed to install the `eslint-import-resolver-alias`.

Here the full working solution for `"vite": "^7.0.3"` and `"eslint": "^9.31.0"` with `"react": "^19.1.0"`

```js
import babelParser from '@babel/eslint-parser';
import js from '@eslint/js';
import importPlugin from 'eslint-plugin-import';
import { defineConfig } from 'eslint/config';
import globals from 'globals';

export default defineConfig([
  importPlugin.flatConfigs.recommended,
  {
    files: ['**/*.{js,jsx,mjs,cjs,ts,tsx}'],
    plugins: {
      ...
    },
    settings: {
      react: {
        version: 'detect',
      },
      'import/resolver': {
        node: {
          paths: ['src'],
          extensions: ['.js', '.jsx', '.tsx', '.ts'],
        },
        alias: {
          map: [
            ['@components', './src/components/'],
           ...
          ],
          extensions: ['.js', '.jsx', '.tsx', '.ts'],
        },
      },
    },
    // Language options are the options for the parser.
    languageOptions: {
      parser: babelParser, 
      parserOptions: {
        requireConfigFile: false,
        babelOptions: {
          presets: ['@babel/preset-react'],
        },
       ecmaFeatures: {
          jsx: true,
        },
        ecmaVersion: 'latest',
        sourceType: 'module',
      },
      globals: {
        ...globals.browser,
        ...globals.node,
      },
    },
    rules: {
      ...
     }
  },
]);
```
